### PR TITLE
feat!: event API

### PR DESCRIPTION
**Pull request by xiaoshihou514**
Add event API to provide control to guard autocmds.

- [x] unify `Guard info` with checkhealth
- [x] option for controlling debounce timer
- [x] `events` field in tool config
- [x] `auto_lint` option
- [x] `Guard lint`
- [x] `Guard disable` -> `Guard disable-fmt`
- [x] `Guard enable` -> `Guard enable-fmt`
- [x] `Guard disable-lint`
- [x] `Guard enable-lint`
- [x] Tests for linting functionality
- [x] Error reporting for linter
- [x] Other minor fixes and refactors
- [x] ! remove `setup`

### CODE REVIEW

File: lua/guard/api.lua [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+---@param bufnr number?
+function M.disable_lint(bufnr)
+  local aus = events.get_lint_autocmds(bufnr or api.nvim_get_current_buf())
+  vim.iter(aus):each(function(au)
+    api.nvim_del_autocmd(au.id)
+  end)
+end
+
+---Show guard info for current buffer
+function M.info()
```
Comment (glepnir) [inline lua/guard/api.lua:53]: we have health.lua with `checkhealth`

Comment (xiaoshihou514) [reply lua/guard/api.lua:53]: Idea is that this is for buf local and checkhealth for global.

**Comment by 231tr0n** [comment]
Apologies for commenting here but are we providing ``lint_on_save`` option similar to ``format_on_save``.

**Comment by xiaoshihou514** [comment]
@231tr0n The goal is to give full control over on what events what should happen.

**Comment by 231tr0n** [comment]
> @231tr0n The goal is to give full control over on what events what should happen.

Ohh so lint_on_save, fmt_on_save and save_on_fmt are no more a thing is it?

**Comment by xiaoshihou514** [comment]
I think I will keep the existing options as nice deafults.

**Comment by 231tr0n** [comment]
> I think I will keep the existing options as nice deafults.

Just a suggestion but I think it would be better to probably remove these options and let users decide when to format and lint and just provide `Guard fmt` and `Guard lint` which they can use anywhere they wish to.

**Comment by xiaoshihou514** [comment]
Yeah I guess it makes sense.

Will call it `auto_lint` since it's not exactly lint on save.

**Comment by xiaoshihou514** [comment]
Reminder to self:
- individual event handling
- check au attach needs to handle custom events

**Comment by xiaoshihou514** [comment]
Need to rewrite the whole event module, will take a while to land

File: lua/guard/api.lua (hunk 1 above)
Comment (glepnir) [reply lua/guard/api.lua:66]: Don't add extra area. It's useless. CheckHealth is the only interface here

File: lua/guard/events.lua [hunk 2]
```diff
@@ -7,35 +7,174 @@ local iter = vim.iter
...
+
+local debounce_timer = nil
+local function debounced_lint(opt)
+  if debounce_timer then
+    debounce_timer:stop()
+    debounce_timer = nil
+  end
+  ---@diagnostic disable-next-line: undefined-field
+  debounce_timer = assert(uv.new_timer()) --[[uv_timer_t]]
+  debounce_timer:start(500, 0, function()
```
Comment (glepnir) [inline lua/guard/events.lua:21]: it's better have a way to control this hardcord time value.

File: lua/guard/filetype.lua [hunk 3]
```diff
@@ -88,17 +94,20 @@ local function box(ft)
...
         M[it] = box(it)
         M[it].linter = self.linter
       end
-      events.register_lint(it, evs)
+
+      if type(config) == 'table' and config.events then
+        -- use user's custom events
+        events.lint_attach_custom(it, config)
+      else
+        events.lint_watch_ft(it, evs)
```
Comment (glepnir) [inline lua/guard/filetype.lua:108]: what does lint_watch_ft do ? for lint we just create ran lint func in callback of autocmds which provide by lint

File: lua/guard/events.lua (hunk 2 above)
Comment (xiaoshihou514) [reply lua/guard/events.lua:21]: How about option `auto_lint_freq`?

Comment (glepnir) [reply lua/guard/events.lua:21]: `lint_interval` would be better

File: lua/guard/filetype.lua (hunk 3 above)
Comment (glepnir) [reply lua/guard/filetype.lua:108]: name should be improved...it make confused. `lint_on_filetype` keep same with autocmd event name

Comment (xiaoshihou514) [reply lua/guard/filetype.lua:108]: sure

File: lua/guard/api.lua (hunk 1 above)
Comment (xiaoshihou514) [reply lua/guard/api.lua:66]: I guess it makes sense to unify both.

File: lua/guard/events.lua [hunk 4]
```diff
@@ -172,7 +172,7 @@ end
 
 ---@param ft string
 ---@param formatters FmtConfig[]
-function M.fmt_watch_ft(ft, formatters)
+function M.fmt_on_ft(ft, formatters)
```
Comment (glepnir) [inline lua/guard/events.lua:175]:
```suggestion
function M.fmt_on_filetype(ft, formatters)
```

File: lua/guard/events.lua [hunk 5]
```diff
@@ -208,7 +208,7 @@ function M.maybe_default_to_lsp(config, ft, buf)
         pattern = ft,
       }) == 0
     then
-      M.fmt_watch_ft(ft, config.formatter)
+      M.fmt_on_ft(ft, config.formatter)
```
Comment (glepnir) [inline lua/guard/events.lua:211]:
```suggestion
      M.fmt_on_filetype(ft, config.formatter)
```

File: lua/guard/events.lua [hunk 6]
```diff
@@ -234,7 +234,7 @@ end
 
 ---@param ft string
 ---@param events string[]
-function M.lint_watch_ft(ft, events)
+function M.lint_on_ft(ft, events)
```
Comment (glepnir) [inline lua/guard/events.lua:237]:
```suggestion
function M.lint_on_filetype(ft, events)
```

File: lua/guard/filetype.lua [hunk 7]
```diff
@@ -78,7 +78,7 @@ local function box(ft)
         -- use user's custom events
         events.fmt_attach_custom(it, config.events)
       else
-        events.fmt_watch_ft(it, self.formatter)
+        events.fmt_on_ft(it, self.formatter)
```
Comment (glepnir) [inline lua/guard/filetype.lua:81]:
```suggestion
        events.fmt_on_filetype(it, self.formatter)
```

File: lua/guard/filetype.lua [hunk 8]
```diff
@@ -105,7 +105,7 @@ local function box(ft)
         -- use user's custom events
         events.lint_attach_custom(it, config)
       else
-        events.lint_watch_ft(it, evs)
+        events.lint_on_ft(it, evs)
```
Comment (glepnir) [inline lua/guard/filetype.lua:108]:
```suggestion
        events.lint_on_filetype(it, evs)
```

File: lua/guard/events.lua (hunk 4 above)
Comment (glepnir) [reply lua/guard/events.lua:175]: same with event name will more clear

**Pull request merged**